[DOCS] Document APIs related to vertex attributes

aidinabedi opened the pull request:
**The description is updated to reflect the latest changes based on feedback.**

The PR includes the following:
- Document property `pc.VertexIterator#element`.
- Document `pc.VertexIteratorAccessor` and its methods.

I confirm I have signed the [Contributor License Agreement](https://docs.google.com/a/playcanvas.com/forms/d/1Ih69zQfJG-QDLIEpHr6CsaAs6fPORNOVnMv5nuo0cjk/viewform).

aidinabedi commented:
@willeastcott Feel I need to add some more explanation here since this PR includes code changes which I **strongly** want to avoid when my objective is to fix typings and documentation.

The biggest challenge I had with this PR was to document `pc.VertexFormat#elements` properly, since it shares a lot with parameter `description` (in the `pc.VertexFormat` constructor). Basically, I had two options:

1. Either document `pc.VertexFormat#elements` as-is and duplicate lots of existing documentation from the big object parameter `description` (in the `pc.VertexFormat` constructor).

2. Or rename a few properties in the `pc.VertexFormat#elements` object and share lots of jsdoc documentation thus avoiding duplication. This has also the side benefit of unifying the two objects `pc.VertexFormat#elements` and parameter `description` which improves API consistency and making it more intuitive.

I went with option 2 because I believe duplication will not only hurt readability and maintainability but also eventually lead to de-sync and confusing code documentation.

willeastcott commented:
Just wanted to drop a quick comment in here. I've scanned the changes here and my reflex reaction is that I suspect this has the potential to break the Editor and certain PlayCanvas apps that leverage this part of the API. I'd have to do a more thorough investigation to weigh up the risks around that. But in short, I can't merge this as fast as your other PRs from yesterday.

aidinabedi commented on src/graphics/vertex-format.js:
```diff
@@ -80,15 +132,19 @@ Object.assign(pc, function () {
+                name: elementDesc.semantic, // backwards compatibility
+                dataType: elementDesc.type, // backwards compatibility
+                numComponents: elementDesc.components // backwards compatibility
             };
```
@willeastcott I appreciate your comment. But please note that the PR still has full backwards-compatibility. Neither the editor nor any app should be affected. The code changes I've made is basically introducing new properties (`semantic`, `type`, `components`) in the `pc.VertexFormat#elements` object to increase consistency and avoid duplication, but at the same time keep the old ones (`name`, `dataType`, `numComponents`) for backwards-compatibility.

All other code changes are just updating references to use the new properties, since they now have documentation. The old properties can still be used and reflect the exact same data. If you look through all references, you will note that these properties are also read-only.

slimbuck replied:
Could make name, dataType and numComponents properties instead?

aidinabedi replied:
@slimbuck Do you mean renaming the other way around?

I think the new properties (semantic, type, components) is shorter and cleaner in general. But sure, that is also possible. Renaming properties in the constructor parameter will likely affect a lot more application code out there since it’s officially documented and has TS declarations (unlike elements). And even thought the API will be backwards-compatible in run-time, more devs will probably be greeted with warnings.

aidinabedi replied:
@slimbuck Do you mean renaming the other way around?

I think the new properties (semantic, type, components) is shorter and cleaner in general. But sure, that is also possible. Renaming properties in the constructor parameter will likely affect a lot more application code out there since it’s officially documented and has TS declarations (unlike elements). And even thought the API will be backwards-compatible in run-time, more devs will probably be greeted with warnings.

aidinabedi commented:
@willeastcott If you are wary about the code changes, I could go with option 1 as described in [my previous comment](https://github.com/playcanvas/engine/pull/1898#issuecomment-593093483):

1. Keep `pc.VertexFormat#elements` as-is and duplicate all the documentation from the object type parameter `description` (in the `pc.VertexFormat` constructor), but use `pc.VertexFormat#elements`  specific property names.

aidinabedi commented:
@willeastcott I've removed all code changes ([by going with option 1](https://github.com/playcanvas/engine/pull/1898#issuecomment-596929377)) in the hopes of getting this PR to pass.

willeastcott commented on src/graphics/vertex-format.js:
```diff
@@ -49,10 +57,101 @@ Object.assign(pc, function () {
+     * a {@link pc.VertexBuffer}.
+     * @description Returns a new pc.VertexFormat object.
+     * @param {pc.GraphicsDevice} graphicsDevice - The graphics device used to manage this vertex format.
+     * @param {pc.VertexAttributeDescription[]} description - An array of vertex attribute descriptions.
```
I'm a bit confused here. If the `pc.VertexFormat` constructor takes an array of `pc.VertexAttributeDescription` objects, then I would expect a PlayCanvas code example to be something like:

```
var position = new pc.VertexAttributeDescription(pc.SEMANTIC_POSITION, 3, pc.TYPE_FLOAT32);
var uv = new pc.VertexAttributeDescription(pc.SEMANTIC_TEXCOORD0, 2, pc.TYPE_FLOAT32);
var color = new pc.VertexAttributeDescription(pc.SEMANTIC_COLOR, 4, pc.TYPE_UINT8. true);
var vertexFormat = new pc.VertexFormat(graphicsDevice, [ position, uv, color ]);
```

But it's still just using vanilla Object instances, right?

aidinabedi replied:
Both `pc.VertexAttributeDescription` and `pc.VertexAttributeElement` are not classes but interfaces. Meaning they are primarily used for documenting (the properties of each object) and provide typings for TypeScript. The interfaces don't actually need to be referenced by any user code.

willeastcott replied:
But how will the user know the difference when reading the API ref?

aidinabedi replied:
The description of each begins with saying it's an interface. However, you have a point, the API jsdoc template doesn't clearly differentiate between classes and interfaces. So I've added a proper constructor to each interface, with all relevant parameters. So that if a dev "accidentally" uses one of the interfaces as a class it will still work as expected.

aidinabedi replied:
@willeastcott Please let me know if you have any remaining concerns.

willeastcott replied:
Sorry for the delay in responding. I hope I don't come across as annoying, but I'm still not getting why this PR is useful. I just did a test with `tsc` and it compiles current `pc.VertexFormat` constructor calls just fine. As of today, the second param for the constructor is nice and succinct, using a simple array of JS objects. This PR effectively introduces a second valid way of calling the constructor (which I feel is confusing for users reading the API ref and browsing existing project sourcebases). This new way of calling the constructor is verbose (`new pc.VertexAttributeDescription` repeated for potentially several attributes). Plus, the PR adds 230 lines to the engine codebase - and I'm struggling to see the benefit.

aidinabedi replied:
@willeastcott Not annoying at all :) I fully understand why it may not be obvious. Let me explain:

If you look closely at the definition of the `pc.VertexFormat` constructor in `playcanvas.d.ts` you will see that the `description` parameter type is incorrectly interpreted as an array of unknown `object`, instead of an array of a well-defined type (i.e. `{ semantic,  components, type, normalize }`):

![image](https://user-images.githubusercontent.com/255073/77848188-66792c00-71c3-11ea-9cc2-66db6d7a3f86.png)


This is because current jsdoc comments use `description[].xxx` to define the properties of the type. It would be fine if `description` was not an array, but a single element. Unfortunately, `tsd-jsdoc` only supports the later case. So either we add support for it in `tsd-jsdoc` which is a lot more complex and uncertain endeavor or we simply move out the jsdoc comments into a separate type, i.e. `pc.VertexAttributeDescription`. Because in the current state, `description` has no proper typings. The later solution has the side benefit of allowing developer code in TypeScript to also reference the element type without having to redefine it (when for example passing on the parameter to PlayCanvas).

willeastcott replied:
So basically this is a matter of getting stricter type checking for Typescript developers when specifying a vertex format description at the cost of adding complexity to the codebase and the API reference. It's not a straightforward win IMHO. I'm actually very sensitive to adding new classes to the sidebar of the API reference manual. It's already massively long and if anything, I want to simplify it. And my problem about it confusing developers is still valid. By adding it to the API ref, we're saying "Yeah, you _can_ also do it this way. You don't have to. But this class is actually just here to make type checking more rigorous". It may well be a lot harder to get this addressed in tsd-jsdoc but that _is_ where the problem lies - not the PlayCanvas Engine API.

aidinabedi replied:
@willeastcott Fair enough. I respect your desire to keep the API classes to a minimum.

What about `VertexAttributeElement`? Unlike parameter `description`, `pc.VertexFormat#elements` is a property and there is [no documented valid way in jsdoc to define array-element-properties using `@property`](https://jsdoc.app/tags-property.html), similar to `description[].xxx`.

Additionally, constructor `VertexIteratorAccessor` takes a `VertexAttributeElement`, which means without this new type we would have to "embed" an anonymous definition (equivalent to `VertexAttributeElement`) twice in two different files and remember to keep them in sync, instead of having everything in one place.

aidinabedi commented:
@willeastcott Any updates?

aidinabedi commented:
Hello @willeastcott,

I've reverted all jsdoc for the `description` param in the `VertexFormat` constructor and removed `VertexAttributeDescription`, as discussed.

Also simplified `VertexAttributeElement` by removing the constructor params (since it's never meant to be constructed by user code). And as mentioned in a [previous comment](https://github.com/playcanvas/engine/pull/1898#discussion_r399804598), I still think there is a stronger need for `VertexAttributeElement` (because the object is shared among multiple engine classes).

Let me know if you have any more comments or thoughts.

willeastcott commented:
Let me put my concern another way. After this PR is merged, I would consult the API ref and assume I can do this:

```
pc.app.root.findComponent('model').meshInstances[0].mesh.vertexBuffer.format.elements[0] instanceof pc.VertexAttributeElement
```

But that would return false.

aidinabedi commented:
@willeastcott I'm all for it (was just trying to avoid code changes this time around). `VertexAttributeElement` is now changed to a regular class and instantiated inside `VertexFormat`. The behavior and properties are all the same.

aidinabedi commented:
@willeastcott Any chance we could merge or close this PR?

I have a subtle feeling you are not the biggest fan of it :)

aidinabedi commented:
@willeastcott Would it help if I just removed `VertexAttributeElement`. I'm happy to do that if that is what it takes. That's probably what you've been trying to say to me this whole time :)

willeastcott commented:
Yeah, sorry this PR is still in limbo. 😞 

Actually, it's still `pc.VertexAttributeElement` that I'm not keen on. I think I've mentioned before, this PR is documented a new PlayCanvas class just to circumvent a deficiency in the JSDoc -> TSD toolchain. Sure, it's harder to fix the issue there, but that's where the problem lies, not PlayCanvas' API.

As for `pc.VertexIteratorAccessor`, that method of iterating vertices in vertex buffers is incredibly inefficient. It's pretty much legacy. We're in the process of developing a new API for fast creation and editing of procedural meshes: https://github.com/playcanvas/engine/pull/1960  This part of the API will be even more irrelevant once we're done with that. I can imagine we might deprecate entirely sometime soon. I feel polishing these APIs, adding more public classes etc, is just encouraging developers to use it, when actually we want to do the opposite.

aidinabedi commented:
@willeastcott Thank you for responding.

I know you mentioned before that you're not keen on new classes just to fix typings. [I responded by trying to argue that `pc.VertexAttributeElement` is different](https://github.com/playcanvas/engine/pull/1898#discussion_r399804598), but obviously failed :)

Anyhow, I hear you. I agree, iterating vertices in a interleaved vertex buffer is inefficient. The thing is, that interleaved data still [considered to have better rendering performance](https://developer.mozilla.org/en-US/docs/Web/API/WebGLRenderingContext/vertexAttribPointer#Description). And from what I've seen in #1960, that PR will not provide a complete substitute for creating and accessing a interleaved buffer. And if a developer only needs to update/create a "procedural" mesh once, they should still have the option of choosing these "legacy" APIs for highest draw performance.

Additionally, as you know we're trying to write an glTF exporter for PlayCanvas. And having proper access to APIs like `VertexIteratorAccessor` is necessary to export vertex data efficiently without copying each attribute separately and re-creating the interleaved buffer ourselves in glTF.

aidinabedi commented:
Alright, @willeastcott.

I've removed the new class `VertexAttributeElement`. This PR now only improves documentation for existing classes. No new classes just to circumvent a deficiency in `tsd-jsdoc`. PR description is updated.

I'm just making one existing class `pc.VertexIteratorAccessor` public, which is already used by devs and exposed by [`pc.VertexIterator.html#element`](https://developer.playcanvas.com/en/api/pc.VertexIterator.html#element), This PR only improves upon that existing documentation. **Please consider it.** Proper documentation and type-safety and avoids a lot of unnecessary bugs.

Additionally, [as mentioned before, the new procedural mesh API won't completely replace the need for `pc.VertexIteratorAccessor`](https://github.com/playcanvas/engine/pull/1898#issuecomment-609057522).

willeastcott commented:
This looks in good shape now. But I have merge #1960 first - a word of warning: there will be some merge conflicts after that. But hopefully they'll be easy to resolve.

willeastcott commented on src/graphics/vertex-iterator.js:
```diff
@@ -46,8 +46,9 @@ Object.assign(pc, function () {
+     * @param {pc.VertexFormat} vertexFormat - A vertex format that defines the layout of vertex data inside the buffer.
      */
-    function VertexIteratorAccessor(buffer, vertexElement) {
+    function VertexIteratorAccessor(buffer, vertexElement, vertexFormat) {
```
Why are you passing `vertexFormat` here? I can't see it being used in the body of the constructor.

willeastcott commented:
Thanks for your patience on this one, @aidinabedi. Feels great to finally merge it. :tada:
